### PR TITLE
Introduce pre- and post-interactions for RFQ

### PR DESCRIPTION
**Pull request by k06a**

### CODE REVIEW

File: contracts/libraries/ConstraintsLib.sol [hunk 1]
```diff
@@ -14,6 +14,8 @@ library ConstraintsLib {
     uint256 private constant _SERIES_MASK = type(uint8).max;
     uint256 private constant _NO_PARIAL_FILLS_FLAG = 1 << 255;
     uint256 private constant _NO_IMPROVE_RATE_FLAG = 1 << 254;
+    uint256 private constant _PRE_INTERACTION_CALL_FLAG = 1 << 253;
+    uint256 private constant _POST_INTERACTION_CALL_FLAG = 1 << 252;
```
Comment (typicalbuster) [inline contracts/libraries/ConstraintsLib.sol:18]: Not sure if its necessary though. If user signed order, which includes interaction data then its itself means that it requires interaction call. Processing order without this interaction data will ofc fail.

Comment (k06a) [reply contracts/libraries/ConstraintsLib.sol:18]: Yes, but we need this flags to indicate if order creator want/expect this calls to happen.

**Review by typicalbuster — APPROVED** [review]

**Comment by typicalbuster** [comment]
@k06a need to fix built

**Comment by k06a** [comment]
@typicalbuster I am considering passing whole `order` structure into this interactions, this will be much easier.

**Comment by k06a** [comment]
@typicalbuster this could be even cheaper in terms of gas if we would use assembly for these calls and `calldatacopy` whole `order`.

**Comment by typicalbuster** [comment]
> @typicalbuster this could be even cheaper in terms of gas if we would use assembly for these calls and `calldatacopy` whole `order`.

@k06a  Why? For ETH RFQ orders you just need to encode (target, asset) and for normal only (target).

**Comment by k06a** [comment]
> > @typicalbuster this could be even cheaper in terms of gas if we would use assembly for these calls and `calldatacopy` whole `order`.
> 
> @k06a Why? For ETH RFQ orders you just need to encode (target, asset) and for normal only (target).

Solidity compiler is not very good in encoding `calldata` for the external call, using assembly is usually a lot cheaper

**Pull request merged**